fen.go

```diff
@@ -9,30 +9,285 @@ package fsnotify
 
 import (
 	"errors"
+	"fmt"
+	"golang.org/x/sys/unix"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 )
 
 // Watcher watches a set of files, delivering events to a channel.
 type Watcher struct {
 	Events chan Event
 	Errors chan error
+
+	port *unix.EventPort
+
+	done     chan struct{} // Channel for sending a "quit message" to the reader goroutine
+	doneResp chan struct{} // Channel to respond to Close
 }
 
 // NewWatcher establishes a new watcher with the underlying OS and begins waiting for events.
 func NewWatcher() (*Watcher, error) {
-	return nil, errors.New("FEN based watcher not yet supported for fsnotify\n")
+	var err error
+
+	w := new(Watcher)
+	w.Events = make(chan Event)
+	w.Errors = make(chan error)
+	w.port, err = unix.NewEventPort()
+	if err != nil {
+		return nil, err
+	}
+	w.done = make(chan struct{})
+	w.doneResp = make(chan struct{})
+
+	go w.readEvents()
+	return w, nil
+}
+
+// sendEvent attempts to send an event to the user, returning true if the event
+// was put in the channel successfully and false if the watcher has been closed.
+func (w *Watcher) sendEvent(e Event) (sent bool) {
+	select {
+	case w.Events <- e:
+		return true
+	case <-w.done:
+		return false
+	}
+}
+
+// sendError attempts to send an error to the user, returning true if the error
+// was put in the channel successfully and false if the watcher has been closed.
+func (w *Watcher) sendError(err error) (sent bool) {
+	select {
+	case w.Errors <- err:
+		return true
+	case <-w.done:
+		return false
+	}
+}
+
+func (w *Watcher) isClosed() bool {
+	select {
+	case <-w.done:
+		return true
+	default:
+		return false
+	}
 }
 
 // Close removes all watches and closes the events channel.
 func (w *Watcher) Close() error {
+	if w.isClosed() {
+		return nil
+	}
+	close(w.done)
+	w.port.Close()
+	<-w.doneResp
 	return nil
 }
 
 // Add starts watching the named file or directory (non-recursively).
 func (w *Watcher) Add(name string) error {
-	return nil
+	if w.isClosed() {
+		return errors.New("FEN watcher already closed")
+	}
+	if w.port.PathIsWatched(name) {
+		return nil
+	}
+	stat, err := os.Stat(name)
```

Comment:
When this function gets renamed to AddRaw as noted above, this call will need switch to `os.Lstat` to not follow symlinks.

Reply:
Good catch. Are there uses of Stat on other OSes that should be changed to Lstat in #289?

Reply:
I haven't looked. I *think* the use of Stat here is somewhat specific to how file watching is done with Event Ports (where you pass in the stat struct rather than just a path.)

My plan is to add some tests that confirm that Lstat is needed in the other two locations as well (under the assumption that `Add` is fully in charge of the dereferencing of a symlink) and see if they pass or fail on the other operating systems.

It occurs to me that I suddenly have questions about how Remove is supposed to work if a path was watched with Add vs with AddRaw. Probably need some more tests around that too...

Reply:
I did look. No other implementation uses os.Stat. Concerns moved to the freshly opened issue and PR.

Reply:
Unresolving this to draw the attention of other reviewers. I *think* this is fine and correct, but it's worth close attention.